packages/Meshfree/src/DTK_DetailsSVDImpl.hpp

```diff
@@ -0,0 +1,304 @@
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void svd_2x2( const matrix_2x2_type A, matrix_2x2_type &U,
+                  matrix_2x2_type &E, matrix_2x2_type &V ) const
+    {
+        matrix_2x2_type At, AAt, AtA;
+        trans_2x2( A, At );
+        mult_2x2( A, At, AAt );
+        mult_2x2( At, A, AtA );
+
+        // Find U such that U*A*A’*U’ = diag
+        auto phi = 0.5 * atan2( AAt[0][1] + AAt[1][0], AAt[0][0] - AAt[1][1] );
+        auto cphi = cos( phi );
+        auto sphi = sin( phi );
+
+        U = {{{{cphi, -sphi}}, {{sphi, cphi}}}};
+
+        // Find W such that W’*A’*A*W = diag
+        auto theta =
+            0.5 * atan2( AtA[0][1] + AtA[1][0], AtA[0][0] - AtA[1][1] );
+        auto ctheta = cos( theta );
+        auto stheta = sin( theta );
+        matrix_2x2_type W = {{{{ctheta, -stheta}}, {{stheta, ctheta}}}};
+
+        // Find the singular values from U
+        auto sum = AAt[0][0] + AAt[1][1];
+        auto dif = sqrt( ( AAt[0][0] - AAt[1][1] ) * ( AAt[0][0] - AAt[1][1] ) +
+                         4 * AAt[0][1] * AAt[1][0] );
+        E = {{{{sqrt( 0.5 * ( sum + dif ) ), 0}},
+              {{0, sqrt( 0.5 * ( sum - dif ) )}}}};
+
+        // Find the correction matrix for the right side (S = U'*A*W)
+        matrix_2x2_type Ut, AW, S;
+        mult_2x2( A, W, AW );
+        trans_2x2( U, Ut );
+        mult_2x2( Ut, AW, S );
+
+        // We need sgn_plus here to work with singular systems. Using the
+        // regular sgn will produce a singular C which would lead to singular
+        // V.
+        matrix_2x2_type C = {{{{double( sgn_plus( S[0][0] ) ), 0.0}},
+                              {{0.0, double( sgn_plus( S[1][1] ) )}}}};
+
+        mult_2x2( W, C, V );
+
+        // assert(check_svd(A, U, E, V) < 1e-14);
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void argmax_off_diagonal( typename shared_matrix::const_type A, int &p,
+                              int &q ) const
+    {
+        const auto n = A.extent_int( 0 );
+
+        p = q = -1;
+        double max = -1;
+
+        for ( int i = 0; i < n; i++ )
+            for ( int j = 0; j < n; j++ )
+                if ( i != j && std::abs( A( i, j ) ) > max )
+                {
+                    p = i;
+                    q = j;
+                    max = std::abs( A( i, j ) );
+                }
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    double norm_F_wo_diag( typename shared_matrix::const_type A ) const
+    {
+        const auto n = A.extent_int( 0 );
+
+        double norm = 0.0;
+        for ( int i = 0; i < n; i++ )
+            for ( int j = 0; j < n; j++ )
+                norm += ( ( i != j ) ? A( i, j ) * A( i, j ) : 0 );
+
+        return std::sqrt( norm );
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(
+        const typename Kokkos::TeamPolicy<ExecutionSpace>::member_type &thread,
+        size_t &num_underdetermined ) const
+    {
+        int matrix_id = thread.league_rank();
```

Comment:
So, right now, this means that the full team works on a single matrix. If we wanted to have each thread work on a separate matrix, we could have replaced this by `matrix_id = thread.league_rank()*thread.eam_size() + thread.team_rank()`.
And I think it may be better to switch to `team_member` instead of `thread` as it is a bit misleading .

Reply:
so are we repeating the computation for every thread in the league?

Reply:
We would if we used more than 1 thread in a team. But current `invertMoments` has `TeamPlicy(num_matrices,1)`.

Reply:
got it - so in cuda does this mean only 1 thread in a warp is working or in fact all threads in a warp will do independent work?